### DISCUSSION TITLE
websocket-next access WebSocketConnection outside of @OnOpen or @OnMessage methods

### DISCUSSION THREAD

**Discussion by sharpe-developer** [quarkusio/quarkus#39898] [Q&A]
Scenario: a websocket-next server accepts a connection from a client and then needs to stream data to the client within a `Runnable` launched within `@OnOpen`. The  websocket-next `WebSocketConnection` is available in `@OnOpen` or `@OnMessage` methods but appears to be invalid in the `Runnable` context ([refer to Access to the WebSocketConnection](https://quarkus.io/guides/websockets-next-reference#access-to-the-websocketconnection)). How can a websocket-next server access the `WebSocketConnection` to send data to the connected client outside of the `@OnOpen` or `@OnMessage` methods?

**Comment by maxandersen** [comment]
I have the same question - unsure what the migration of jakarta's `.sendText(msg, handler)` actually is as in the handler I need to sometimes access the connection but I only have a Uni....

would be good to document some patterns for it.

**Reply by mkouba** [reply]
> migration of jakarta's `.sendText(msg, handler)` actually is

Which refers to which method?

**Comment by carterventures** [comment]
Bumping this over a year later - is this a supported use case? There doesn't appear to be any way to send messages from outside the websocket endpoint to clients that are connected. OpenConnections as documented is injected as null in other beans so can't be iterated, and there doesn't seem to be any other way to send messages into the websocket context in websockets-next. Are we missing something?

**Reply by mkouba** [reply]
> OpenConnections as documented is injected as null in other beans

@sharpe-developer That should not be the case. It's implemented as a `@Singleton` bean, so it should never be injected as `null`. Could you describe your use case in more detail?

**Reply by sharpe-developer** [reply]
@mkouba I'm sorry, its been almost 2 years and I'm not actively working on this anymore.

If I recall, the issue was as described in my first message. The `websocket-next` based server could not access the `WebSocketConnection`  outside of the `@OnOpen` or `@OnMessage` methods. This made it impossible to send data to a connected client from another context such as a `Runnable` spawned in `@OnOpen`

I ended up using the legacy websocket implementation because it supported my use case.

**Reply by mkouba** [reply]
> @mkouba I'm sorry, its been almost 2 years and I'm not actively working on this anymore.

Ok, no problem.

> 
> If I recall, the issue was as described in my first message. The `websocket-next` based server could not access the `WebSocketConnection` outside of the `@OnOpen` or `@OnMessage` methods. This made it impossible to send data to a connected client from another context such as a `Runnable` spawned in `@OnOpen`

It can access the connection outside the callbacks if it's obtained from an injected `OpenConnections` or accepted as a method param of the callback. The problem appears when you attempt to access the `WebSocketConnection` injected directly in the enpoint (field, constructor, etc.) - this is a technical problem of `@SessionScoped` beans.

> 
> I ended up using the legacy websocket implementation because it supported my use case.

Ok, but keep in mind that this limitation is not there (for next time ;-).

**Reply by sharpe-developer** [reply]
Thank you @mkouba.

I won't be able to test this solution for reasons stated previously. However, others in this thread (@carterventures @maxandersen) posted more recently and may be actively seeking a solution. I suggest working with them directly to test this solution and ensure this issue is resolved by your proposal.

**Reply by mkouba** [reply]
@carterventures @maxandersen WDYT?

**Reply by carterventures** [reply]
Hi @mkouba, thanks for responding - our generalized use case was we want to bridge connections to a web socket endpoint with events / comms from elsewhere in the application. Meaning we want to send messages to various websocket clients based on logic in a separate REST call or events received from separate pipelines.

In the docs it looked like we could inject OpenConnections into a separate bean to do this (during testing we created a bean called "EventService" and @ApplicationScoped it). Calls to a separate rest endpoint would then use the OpenConnections instance to iterate open connections and find those that were applicable, then send arbitrary events to the connection. Our problem was that OpenConnections wasn't being injected into the EventService bean - it was always a null instance.

We might have missed something, but it seemed like the overall process was intended to be pretty simple according to the docs.

**Reply by mkouba** [reply]
> Our problem was that OpenConnections wasn't being injected into the EventService bean - it was always a null instance.

This is really weird. In CDI, a `null` can be only injected for a `@Dependent` producer method (that returns `null`) which is not the case of `OpenConnections`. Could you share a simple reproducer, e.g. a simplified `EventService` class?
